feat: add extension provider docs (#529)
nitrictech/docs: src/pages/reference/providers.mdx

# Providers

> Nitric can declare and interact with cloud features in a way that doesn't couple with the cloud. Nitric providers are the abstraction layer that enables this.

Reviewer: ```suggestion
Nitric can declare and interact with cloud features in a way that is decoupled from any particular cloud provider. Nitric providers are the abstraction layer that enables this.
```

> Providers are individual plugins comprised of two components, the logic which deploys your services and the logic to handle SDK calls at runtime.

Reviewer: ```suggestion
Providers are separate plugins comprised of two components—a Deployment Engine, which automatically deploys your cloud resources and a Runtime Adapter, which handles requests from the SDKs forwarding them to appropriate cloud service APIs.
```

## How it works

> The code you write declares a number of resources, services, and permissions. The following code defines an API named `main` and a bucket named `images`, with permissions to `read` and `write`.

Reviewer: ```suggestion
Applications built with nitric declare resources, services, and permissions directly in the application code. The following example defines an API named `main`, a bucket named `images`, and permissions for the service to `read` and `write` from the bucket.
```

> ```js

Reviewer: ```suggestion
```js {{ tag: 'services\example.js' }}
```

import { api, bucket } from '@nitric/sdk'

const images = bucket('images').allow('read', 'write')
const mainApi = api('main')
```

> The Nitric CLI would take these code declarations and send it to the provider for deployment. The provider will then convert this declaration to actual cloud infrastructure. The specific cloud service that the provider will use for each Nitric resource is listed in the [table below](#standard-providers).

Reviewer: ```suggestion
The choosen language SDKs forward these resource requests to the Nitric CLI, which compiles a graph of the required resources and their connections. The graph of cloud resources is then sent the chosen provider for deployment. The provider convert this declaration graph into any cloud resources it wishes that fulfill the developers intentions, typically using Infrastructure as Code tools like Pulumi or Terraform.

The specific cloud services that NItric's out-of-the-box providers use for each resource is listed in the [table below](#standard-providers).
```

```js
mainApi.get('/', async (ctx) => {
  const imageContents = await images.file('cat.png').read()

  ctx.res.body = imageContents

  return ctx
})
```

> The runtime component will take the abstract calls that you make to the SDK at runtime, like calling `.read()` on a file, and convert that into cloud-specific API calls. This allows your functions to run completely separated from the underlying cloud service.

Reviewer: ```suggestion
The runtime adapter acts as a nitric protocol compliant server, accepting runtime calls made by a nitric client (such as one of the nitric language SDKs), for example calling `.read()` on a file in a bucket. The runtime adapter translates these requests into cloud-specific API requests, typically abstracting common boilerplate such as locating SDK credentials for the cloud specfic SDK. 

This allows your functions to be built, testing and run independent of any underlying cloud service. The end result is code that's faster to write, easier to test and remains portable between cloud services and cloud providers.

It also significantly reduces the burden of writing project specific Infrastructure as Code and ensures the application and its deployment automation don't drift apart as development continues.
```

![Nitric Architecture](/docs/images/guides/custom-provider/custom-provider-architecture.png)

You can find out more about how the Nitric CLI and SDK interact with providers on our [concepts page](/guides/getting-started/concepts).

## Standard Providers

> The providers for AWS, GCP, and Azure are built and maintained by Nitric in our open-source repository on [GitHub](https://github.com/nitrictech/nitric/tree/main/cloud). These providers enable deploying and running your code across AWS, Google Cloud, and Azure. The following is the underlying cloud services that each of the standard providers use:

Reviewer: ```suggestion
Nitric publishes standard providers for AWS, GCP, and Azure, which are available as part of the core open-source repository on [GitHub](https://github.com/nitrictech/nitric/tree/main/cloud). These providers enable deploying and running code across AWS, Google Cloud, and Azure. The following is a description of the underlying cloud services that each of the standard providers use:
```

| [Schedules](/schedules)                | [CloudWatch Event Bridge](./providers/aws/schedules) | [Dapr Binding](./providers/azure/schedules) | [Cloud Scheduler](./providers/gcp/schedules) | Custom                                                                             |
| [Secrets](/secrets)                    | [Secrets Manager](./providers/aws/secrets)           | [Key Vault](./providers/azure/secrets)      | [Secret Manager](./providers/gcp/secrets)    | Custom                                                                             |
| [Storage](/storage#buckets)            | [S3](./providers/aws/storage)                        | [Blob Storage](./providers/azure/storage)   | [Cloud Storage](./providers/gcp/storage)     | SeaweedFS                                                                          |
| Services                               | Lambda                                               | Container Apps                              | CloudRun                                     | Docker                                                                             |

> Our code is completely open-source on our [GitHub](https://github.com/nitrictech/nitric), so you can know exactly how your resources are being deployed and handled at runtime.

Reviewer: ```suggestion
Nitric is open-source and available on [GitHub](https://github.com/nitrictech/nitric), so you can know exactly how your resources are being deployed and handled at runtime.
```

> Our code is completely open-source on our [GitHub](https://github.com/nitrictech/nitric), so you can know exactly how your resources are being deployed and handled at runtime.

Reviewer: ```suggestion
Our code is completely open-source on our [GitHub](https://github.com/nitrictech/nitric), so you can know exactly how your resources are being deployed and handled at runtime.

## Custom Providers

Because nitric providers are plugins you're able to create or extend them to meet your specific deployment requirements. Common reasons for custom provider development include:

- Using a language other than Go to develop a provide (nitric providers can be written in many other languages)
- Using a specific Infratructure as Code engine, e.g. using Terraform or AWS CDK instead of Pulumi
- Changing the way resources are deployed to use a preferred service or meet regulatory or policy requirements within your organization.
- Reusing existing work on infrastructure automation, such using existing Terraform Modules for deployment with Nitric
- Deployments to another cloud provider or on-premises deployments

You can read more about building custom providers [here](/reference/providers/custom)

<Note>
Improving the ease and flexibility of custom provider development is currently a priority for the Nitric team, if you need help or want to request a feature get in touch on [GitHub](https://github.com/nitrictech/nitric) or [Discord](https://nitric.io/chat).
</Note>
```

Reviewer 2: Applied and fixed link to building custom providers
